### PR TITLE
[WIP] AIM Improvements, Yet Again!

### LINKED ISSUE #11996: [0.C 2982] Advanced Inventory crash when moving items
I backtracked through the builds, and this started happening in 2982, up to 2985 (the current version at this time)

The first issue is when you use the advanced inventory management to either A) move any stacking item (batteries, wire, etc.) to another square, or B) move ANY item, stacking or not, from your inventory to the ground. If it's a stacking item, it will auto-input the number for all but ONE of the item in the quantity window, and if you finish the move, they'll move just fine, but the one will obviously be left behind.

The second issue is when you try to move that last item, OR when you try to move any single-stack item out of your inventory. At this point, the quantity window pops up, auto-fills in a 0, and obviously won't move the item after finishing. And if you try to change the 0 to a 1 and make the move, the game crashes with this message:

Assertion failed!
Program: C:\Cataclysm DDA\cataclysmdda\cataclysm-tiles.exe
File: src/advanced_inv.cpp, Line 1381

Expression: amount_to_move > 0

Abort closes the game, retry comes up with a runtime crash, and ignore goes back to the game, with the item unmoved.

Doesn't really make a lot of sense, because there wasn't even anything changed in advanced_inv.cpp in 2982 as far as I can tell XD

(this is my first time making a post here, let me know if you need any more info or if I need to explain it better :D)

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/10981514-0-c-2982-advanced-inventory-crash-when-moving-items?utm_campaign=plugin&utm_content=tracker%2F146201&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F146201&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### LINKED ISSUE #11999: Advanced Inventory item duplication
The way this could be replicated is by having a tab set to [A]ll and moving items from a vehicle (source is VH). If the specific tile is selected instead of [A]ll then it behaves normally.
The other tab doesn't seem to matter.

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/11009170-advanced-inventory-item-duplication?utm_campaign=plugin&utm_content=tracker%2F146201&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F146201&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### LINKED ISSUE #12012: In the Advanced Inventory the Examine function isn't working properly when a pane is set to Worn Items
When Examining items from the [W]orn Items view it shows the description as if the [I]nventory view was actually selected.

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/11077491-in-the-advanced-inventory-the-examine-function-isn-t-working-properly-when-a-pane-is-set-to-worn-items?utm_campaign=plugin&utm_content=tracker%2F146201&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F146201&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### LINKED ISSUE #12016: Advanced INV manager eating items
Win 7 64bit
cata version 24fe0e0 64 bit

Hi all I have an issue

I had a boot full of stuff and I wanted to transfer items from one car to another so used the ADV inventory manager.

Well I was messing about trying to figure it all out using D,V etc but realised all my stuff had gone.

I exited ADV INV manager and looked on the car tile I had all my gear and nothing was there it vanished :(.

No I can promise it was ADV inventory manager but my items where gone I did nothing so its either.

Advanced inventory manager
Auto save

or some other wierd bug :(

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/11131818-advanced-inv-manager-eating-items?utm_campaign=plugin&utm_content=tracker%2F146201&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F146201&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### LINKED ISSUE #12037: Advanced inventory manager - 'move all items' option works only on vehicle storage, not ground
0.C-2992, Windows 7 w/tiles.  In the AIM, select a tile that has items both on the ground and in vehicle storage.  Make sure the ground is selected, then try to move all.  Items will be moved from the vehicle storage instead, and not all of them either, just a couple.

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/11238532-advanced-inventory-manager-move-all-items-option-works-only-on-vehicle-storage-not-ground?utm_campaign=plugin&utm_content=tracker%2F146201&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F146201&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by DavidKeaton**
### [status]

Most, if not all, of the dupe and removal bugs _should be fixed_. _PLEASE_ tell me if they persist in this branch. See the bug list on the bottom for what all I am fixing, and have fixed. Sorry again ya'll. Now I know to do a bit more extensive bug testing!

(p.s. if you need to fastline the bug fixes for dupes and stuff I can make another PR that takes care of the rest afterwards)
- Two vehicle panes at the same time!
- Colored the tab that the [V]ehicle tab is referring to!
- Auto select a vehicle tile if the tile has items, or if both the ground and vehicle tile are empty.
- _Less dupe bugs_ (sigh)

Clears out these fantastic bugs I introduced:
- Fixes #11999
- Fixes #12016
- Fixes #11996
- Fixes #12012
- Fixes #12037

Bug list:
- [x] `AIM_WORN` duplication bug
- [x] `AIM_ALL` duplication bug[s]
- [x] `AIM_WORN` assertion bug
- [ ] `move_all_items()` doing jack shit
- [x] `AIM_ALL` destination bug
- [ ] `AIM_CONTAINER` selection bug
- [x] `AIM_VEHICLE` incorrect area bug
- [x] `AIM_VEHICLE` interface bugs
- [x] `AIM_WORN` examine bug

hth and enjoy! :-)
-Davek

### CODE REVIEW

File: src/advanced_inv.cpp [hunk 1]
```diff
@@ -631,40 +636,26 @@ void advanced_inventory::init()
...
 
-    panes[left].area = AIM_ALL;
-    panes[right].area = AIM_INVENTORY;
-
-    panes[left].sortby = ( advanced_inv_sortby ) uistate.adv_inv_leftsort;
-    panes[right].sortby = ( advanced_inv_sortby ) uistate.adv_inv_rightsort;
-    panes[left].area = ( aim_location ) uistate.adv_inv_leftarea;
-    panes[right].area = ( aim_location ) uistate.adv_inv_rightarea;
-
     bool moved = ( uistate.adv_inv_last_coords.x != g->u.posx() ||
```
Comment (Coolthulhu) [inline src/advanced_inv.cpp]: How about: `uistate.adv_inv_last_coords != g->u.pos3()`

Comment (DavidKeaton) [inline src/advanced_inv.cpp]: minor leftover, all other stuff should be z friendly

File: src/advanced_inv.cpp [hunk 2]
```diff
@@ -631,40 +636,26 @@ void advanced_inventory::init()
...
-            }
-            if(panes[right].area == AIM_VEHICLE) {
-                panes[right].area = AIM_INVENTORY;
-            }
-            // disable loading our data to panes[right]event doing this over again
-            uistate.adv_inv_load_veh = false;
-        }
-    }
+    panes[left].sortby  = (advanced_inv_sortby) static_cast<advanced_inv_sortby>(uistate.adv_inv_leftsort);
+    panes[right].sortby = (advanced_inv_sortby) static_cast<advanced_inv_sortby>(uistate.adv_inv_rightsort);
```
Comment (BevapDin) [inline src/advanced_inv.cpp]: I see, you want to be really, really sure it's an `advanced_inv_sortby` value, so you cast it with a `static_cast` and additionally with a C-style cast. One never knowns what those fancy static_casts return, right? They could return anything, better make sure it's the correct type (-:

**Comment by drbig** [comment]
I obviously didn't read enough, and won't now, so I'll just spit my venom here, based on `0.C-1577-g24fe0e0`:

* Breaks default behaviour that while on vehicle the vehicle tiles are default
* It's impossible to select two different vehicle tiles
* No 'all surrounding area' for vehicle tiles

Also - it's *broken*, steps to reproduce:
* Select 'all surrounding area' while away from vehicle, exit AIM
* Move onto a vehicle tile
* Enter AIM, have fun with infinite items supply from the 'all surrounding area' tiles

Also while I was at it I've tried something else and got:
```
cataclysm-tiles: src/advanced_inv.cpp:1381: void advanced_inventory::display(): Assertion `amount_to_move > 0' failed.
```
no idea what I did exactly... and lost an hour of driving west. :S

I assumed the main idea was to add 'reaching ground tiles while in vehicle + perks', not 'fsck current AIM behaviour + bugs'.

*__END_OF_RANT__*

I hope the above is known and in the pipeline for fixing and that my input here is redundant. :/

**Comment by Rivet-the-Zombie** [comment]
Also it seems that opening the adv. inventory with a vehicle tile preselected while away from any valid vehicles throws a snarky debug message about wishing for a Ferrari.

**Comment by DavidKeaton** [comment]
My intents were not snarky. :( However I am gonna remove the text there anyway as I gotta make some changes.

File: src/advanced_inv.cpp (hunk 2 above)
Comment (DavidKeaton) [inline src/advanced_inv.cpp]: LOL whoops! hehe

File: src/advanced_inv.cpp [hunk 3]
```diff
@@ -2190,45 +2238,49 @@ void advanced_inventory::draw_minimap()
...
     // setup on the square's offsets, as a second safeguard
-    auto &valid_square = squares[squares[sel].offset_to_location()];
+    auto &valid_square = squares[loc];
     squares[AIM_VEHICLE].set_vehicle(valid_square);
     return true;
 }
 
 void advanced_inv_area::set_vehicle(advanced_inv_area &s)
 {
+    if(veh == nullptr) {
```
Comment (BevapDin) [inline src/advanced_inv.cpp]: Well, you call this function from the other `set_vehicle` function on `squares[AIM_VEHICLE]`, which has been initialized (see `advanced_inv_area::init`) to have `veh == nullptr` (the case for `AIM_VEHICLE` does not set `veh` to anything). As a result, this test is usually true, especially when re-entering the AIM.

Comment (DavidKeaton) [inline src/advanced_inv.cpp]: Yes I meant s.veh. Sorry, the last bit of stuff was done in a bit of a rush.

File: src/advanced_inv.cpp [hunk 4]
```diff
@@ -1727,6 +1773,7 @@ void advanced_inventory::remove_item( advanced_inv_listitem &sitem )
         assert( &cont->contents.front() == sitem.it );
         cont->contents.erase( cont->contents.begin() );
     } else if(sitem.area == AIM_VEHICLE || sitem.area == AIM_DRAGGED) {
+        set_vehicle(panes[src], panes[src].veh_area);
         if(s.can_store_in_vehicle()) {
```
Comment (BevapDin) [inline src/advanced_inv.cpp]: This test seems strange: `remove_item` has no return value, nothing to inform the caller whether removing was done or not. The caller will simply assume it has been done, and it's the callers responsibility to assure removing the item is actually possible (otherwise it must not call `remove_item`).

Now this test seems to indicate there is a situation where the item removal is not possible. This will essentially lead to item duplication as the caller will add the item to the target location and afterwards remove the original item (but the removal might fail here).

File: src/advanced_inv.cpp [hunk 5]
```diff
@@ -1746,16 +1793,16 @@ bool advanced_inventory::add_item( aim_location destarea, const item &new_item,
...
-        } else {
-            return g->u.wear_item(&g->u.i_add(new_item));
+        if((rc = g->u.wear_item(&new_item)) == false) {
+            popup(_("You can't wear that!"));
         }
+        return rc;
     } else {
         advanced_inv_area &p = squares[destarea];
+        // set the vehicle area for the destination area
+        if(squares[destarea].can_store_in_vehicle()) {
```
Comment (BevapDin) [inline src/advanced_inv.cpp]: `can_store_in_vehicle` returns true if there is a vehicle there, it does not check whether I actually *want* to put it into the vehicle. Selecting a square that has a vehicle (but not selecting the vehicle storage) will make this condition true. Consequently `set_vehicle` is called, which calls `swap_panes` and causes the "This here be a bug! [NUM_AIM_LOCATIONS] in set_vehicle()" debug message (twice, a second time from `remove_item` for the same reason).

I think you might be better of with a simply bool in `advanced_inventory_pane` and in `advanced_inventory_area` that indicates "use the vehicle storage (if it exists)" instead AIM_VEHICLE, and consequently remove "AIM_VEHICLE" completely.

File: src/advanced_inv.cpp [hunk 6]
```diff
@@ -1362,7 +1383,9 @@ void advanced_inventory::display()
             if( !query_destination( destarea ) ) {
                 continue;
             }
-            if( squares[srcarea].is_same( squares[destarea] ) ) {
+            if( squares[srcarea].is_same( squares[destarea] ) &&
```
Comment (BevapDin) [inline src/advanced_inv.cpp]: This part (the block starting at line 1356|1377 with `action == "MOVE_SINGLE_ITEM") has one big problem:
 the source and target are only defined as `aim_location`, which basically excludes vehicle-to-vehicle transfers (`srcarea == AIM_VEHICLE`, `destarea == AIM_VEHICLE`, huh?)

I see at least two possible solutions:
- Move most of the code into the `advanced_inventory_pane` class and use the members of that class to identify where to move from/to (use `area` *and* `veh_area` and maybe more).
- Add more aim_locations: one for each direction (N, S, W, ...) and ALL, but for vehicle storage, they would act like AIM_VEHICLE but with specific locations. (AIM_VEHICLE does not have a location, doesn't it?)

**Comment by drbig** [comment]
So I've actually written a 400 word piece first... I'll *try* to distil:

Per pane state:
* `AIM_SOURCE = all, ground, vehicle, dragged, container, inventory` - all includes ground and any vehicle
* `AIM_DIRECTION = cardinals, ordinals, below, all` - only valid for all - dragged of the above

Per pane behaviour:
* Select vehicle source when default/last/newly selected direction has vehicle storage, if not select ground (preserve old behaviour = if there is a vehicle storage, it gets selected by default; that is regardless of current source selection, and yes you'll need to hit another button to go back to ground; ground is *additional* feature here)
* The difference between dragged vehicle and vehicle source is that in dragged mode the directional selections are limited to only valid dragged vehicle storage tiles (and by default a dragged vehicle storage tile is selected when switching to that mode) (or at least the latter part)

Questions:
* Should player be able to select ground when at a non-edge vehicle tile? How do you justify even knowing what's below the middle of a 3x3 (10x10...) tank? 'Went out, crawled below?', 'Your vehicle has no solid floor?' And then you're supposed to be moving items between that place any any other possible choice.

TL;DR: This gets complicated even on the purely conceptual level.

**Comment by dvolk** [comment]
Can you do a seperate PR for the two bug fixes? They're pretty critical and this PR seems like it might take a while.

**Comment by DavidKeaton** [comment]
@dvolk I can do that if that is the best course of action. I do, however, have most of this done from what I can tell. (albeit I am sure that doesn't mean much to ya'll now lol)

File: src/advanced_inv.cpp (hunk 6 above)
Comment (DavidKeaton) [inline src/advanced_inv.cpp]: It stores the attributes of the location it is representing.

**Comment by DavidKeaton** [comment]
@drbig I have implemented defaulting to a vehicle storage tile, unless the ground has items and the vehicle storage there has no items. Otherwise, it is default. (if veh has any items, or if no items on either, vehicle storage = selected)

**Comment by Coolthulhu** [comment]
The oldest bug on the list has 5 days and your list still isn't 100%. Does the removal of [WiP] from the name mean it is no longer WiP or only that the list now represents the progress?

Kevin granted me the dev powers and I currently have a lot of free time, so I could mergetest it right now. Should I?

**Comment by tivec** [comment]
@DavidKeaton The last commit, 1d8b14a, will not compile for me. Some errors I am getting seem to relate to a use of undeclared identifier AIM_VEHICLE and comparison with different enumeration types ('aim_location' and 'activity_type'). 

Check this gist for the complete build log:
https://gist.github.com/b6165f78167a31e6785e

**Comment by DavidKeaton** [comment]
@Coolthulhu I need to re-add the [WIP] tag, as I still need to wrinkle out a lot of these changes. Most of the bugs have been knocked out but I still need to fix up a few of the other ones.

@tivec I think the latest commit should be better? I will run through it all to make sure, but I try to do a quick compile test before updating the commit to github.

**Comment by DavidKeaton** [comment]
@Coolthulhu tell you what bud, I am proposing a revert of my prior additions to the master branch. I didn't realize how in depth I had to go, and didn't test well enough.

**Comment by Coolthulhu** [comment]
Last commit was a while ago (almost 2 weeks) and it also fixed some earlier bugs.

You sure you can't patch it up quickly? Get it up to compile, move all items from one place to the other - at least the simple stuff. For now you could totally disable worn item interaction and focus just on restoring all the old behaviors.

**Comment by DavidKeaton** [comment]
okay bud, I can shoot for at least old functionality. :-)

**Comment by Coolthulhu** [comment]
You still on it? 3 days and no commit.
Is it OK if I take what you managed so far and try to finish it? I'd rather not revert everything, but lack of AIM kinda hurts.

**Comment by DavidKeaton** [comment]
yes that's fine. been under the weather, sorry about that. :-(

**Comment by Coolthulhu** [comment]
I have a bit of free time right now, so I'll finish it up and merge it as a bugfix.

**Comment by DavidKeaton** [comment]
thanks m8 :-)

**Comment by Coolthulhu** [comment]
I fixed it up and merged it as a bugfix. Even if my fix is broken, it's still better than not including all your fixes.

The duping that happened when moving items happened before you made different assumptions from the author of pickup/move code: you seem to have assumed that moving a "quantity 1 of rock" will move a single rock (which makes sense).
Author of pickup/move programmed it so that quantity 0 means "move everything", while quantity 1 means "split this rock item into rock with 1 charge and rock with (charges-1) charges".

This happened while moving all items around the map and while picking up a lot of items.

What I didn't fix:
* It is still not possible to move all items under a vehicle cargo. They'll be placed in the cargo or not placed at all.
* It is not possible to pick vehicle tile and the tile under this one at the same time.
* When targeting worn items and moving 2 (or more) items of the same type, only 1 of them will be worn.

**Comment by DavidKeaton** [comment]
that explains a ton, lol. also minor jostling and I can get the interface bit. thanks man :-)

**Comment by Coolthulhu** [comment]
Small request from me: it would be cool if you managed to make "move all" work with source area `AIM_ALL`. Even if it was just enumerating all areas not equal to destination area.
It would be useful for making neat piles out of wrecked cars.

**Comment by DavidKeaton** [comment]
I can agree to that. :-)

**Comment by Coolthulhu** [comment]
Forgot to add: didn't touch the container bug.
